Enable to customize completion commands

suzuki-shunsuke opened the pull request:
<!--
  This template provides some ideas of things to include in your PR description.
  To start, try providing a short summary of your changes in the Title above.
  If a section of the PR template does not apply to this PR, then delete that section.
 -->

## What type of PR is this?

_(REQUIRED)_

<!--
  Delete any of the following that do not apply:
 -->

- feature

## What this PR does / why we need it:

_(REQUIRED)_

<!--
  What goal is this change working towards?
  Provide a bullet pointed summary of how each file was changed.
  Briefly explain any decisions you made with respect to the changes.
  Include anything here that you didn't include in *Release Notes*
  above, such as changes to CI or changes to internal methods.
-->

To customize completion command's fields such as `Hidden`, `Usage`, and `Description`.

## Which issue(s) this PR fixes:

_(REQUIRED)_

- https://github.com/urfave/cli/discussions/2102

<!--
If this PR fixes one of more issues, list them here.
One line each, like so:

Fixes #123
Fixes #39
-->

## Special notes for your reviewer:

_(fill-in or delete this section)_

<!--
   Is there any particular feedback you would / wouldn't like?
   Which parts of the code should reviewers focus on?
-->

## Testing

_(fill-in or delete this section)_

<!--
  Describe how you tested this change.
-->

```go
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	"github.com/urfave/cli/v3"
)

func main() {
	cmd := &cli.Command{
		Name: "greet",
		// EnableShellCompletion is unnecessary
		ConfigureShellCompletionCommand: func(cmd *cli.Command) {
			cmd.Hidden = false
			// cmd.Usage = "shell comletion"
			// cmd.Description = "Generate shell completion scripts"
		},
		Commands: []*cli.Command{
			{
				Name:  "hello",
				Usage: "Say hello",
				Action: func(ctx context.Context, cmd *cli.Command) error {
					fmt.Println("Hello")
					return nil
				},
			},
		},
	}

	if err := cmd.Run(context.Background(), os.Args); err != nil {
		log.Fatal(err)
	}
}
```

```console
$ go run main.go --help           
NAME:
   greet - A new cli application

USAGE:
   greet [global options] [command [command options]]

COMMANDS:
   hello       Say hello
   help, h     Shows a list of commands or help for one command
   completion  Output shell completion script for bash, zsh, fish, or Powershell

GLOBAL OPTIONS:
   --help, -h  show help

$ go run main.go completion --help
NAME:
   greet completion - Output shell completion script for bash, zsh, fish, or Powershell

USAGE:
   greet completion [command [command options]]

DESCRIPTION:
   Output shell completion script for bash, zsh, fish, or Powershell.
   Source the output to enable completion.

   # .bashrc
   source <(greet completion bash)

   # .zshrc
   source <(greet completion zsh)

   # fish
   greet completion fish > ~/.config/fish/completions/greet.fish

   # Powershell
   Output the script to path/to/autocomplete/greet.ps1 an run it.


OPTIONS:
   --help, -h  show help
```

## Release Notes

_(REQUIRED)_
<!--
  If this PR makes user facing changes, please describe them here. This
  description will be copied into the release notes/changelog, whenever the
  next version is released. Keep this section short, and focus on high level
  changes.

  Put your text between the block. To omit notes, use NONE within the block.
-->

```release-note

```

dearchap commented on completion.go:
```diff
@@ -39,10 +40,28 @@ var (
+Output the script to path/to/autocomplete/$COMMAND.ps1 an run it.
+`
+
+func BuildCompletionCommand(appName string) *Command {
```
Why the name change to make it public ?

suzuki-shunsuke replied:
- https://github.com/urfave/cli/discussions/2102

To allow to change the completion command's filelds such as `Hidden`.

dearchap reviewed (commented):
@suzuki-shunsuke we have the docs to show how to invoke this https://cli.urfave.org/v3/examples/completions/shell-completions/

We generally want to keep the cli binary to a minimum so we avoid unnecessary descriptions/etc. I am not against adding the command description as part of the command but I would like others to chime in here

suzuki-shunsuke commented:
> suzuki-shunsuke we have the docs to show how to invoke this https://cli.urfave.org/v3/examples/completions/shell-completions/
> 
> We generally want to keep the cli binary to a minimum so we avoid unnecessary descriptions/etc. I am not against adding the command description as part of the command but I would like others to chime in here

I'm aware the document.
But it doesn't describe how to customize fields such as `Hidden`, `Usage`, and `Description` of `completion` command.
Now `buildCompletionCommand` is private, so there is no way to customize them unless overriding the command itself.

- https://github.com/urfave/cli/discussions/2102

I set the default usage and description to make it user friendly, but if you are against, it's okay to remove them.
Users can set them explicitly.

dearchap commented:
@suzuki-shunsuke do you want to update the docs under docs/v3/completions to include this as well ? That way I can push docs out to server once we merge this in

suzuki-shunsuke commented:
> suzuki-shunsuke do you want to update the docs under docs/v3/completions to include this as well ? That way I can push docs out to server once we merge this in

Sure. Updated. 23f8f563b32adb2f6000b8d17f91883ff3aa01c2

dearchap commented:
@suzuki-shunsuke urfave/cli is declarative in nature. So instead of exporting the completion command we should add a flag saying whether the completion command should be hidden or not. So usage would be 

```
  cmd := &Command{
      Name: "foo",
      ShowCompletionCommand: true
....
```

Advantage of this method is that we dont need to export the completion command. Disadvantage is that we can change only the hidden value for command but not usage or description.

suzuki-shunsuke commented:
How about this?
https://github.com/urfave/cli/commit/98bb3899329c95b3862a4b98aa2fc70d4700af7e

dearchap commented:
@suzuki-shunsuke You have to run "make v3approve" since the public API has changed

suzuki-shunsuke commented:
I updated this pull request and added a field `ConfigureShellCompletionCommand` to *cli.Command.

```go
// ConfigureShellCompletionCommand is a function to configure a shell completion command
type ConfigureShellCompletionCommand func(*Command)
```
